### DISCUSSION TITLE
In need of dynamic form generation solution.

### DISCUSSION THREAD

**Discussion by karlkras** [react-hook-form/react-hook-form#7395] [Q&A] [answered]
Hey folks, so I'm coming to the end of my rope on this. I have a requirement to build out forms dynamically from, e.g.,, json input.
This would be in a react/typescript environment.
I've come close a few time, only to be stopped by one thing or another. The main sticking point I have is the type declarations imposed by typescript does not lend itself to the runtime dynamic construction that I require. Consider this last one, I was hoping to convert [from an existing example](https://stackblitz.com/edit/reusable-rhf-ts-pt6?file=src/components/organisms/registration-form.tsx) but when I got to the point of working out the registration process, I hit this:
```
import React, {FC} from 'react';
import {useForm} from 'react-hook-form';

export type RegistrationFormFields = {
    firstName: string;
    lastName: string;
    email: string;
    bio?: string;
};
export const RegistrationForm: FC = () => {
    const {
        register,
        handleSubmit,
        formState: {errors},
    } = useForm<RegistrationFormFields>();
```
which has been pretty typical on all of my attempts. The static nature of how typescript imposes types get's in the way.
I've found a few examples of "vanilla" react solutions... in fact I've rolled my own in the past, but the addition of typescript has created a block I can't seem to work around and am very surprised that I can find literally nothing that fully addresses this need.
Can someone please point me in a direction that I can gain some footing on this?

**Comment by bluebill1049** [comment, marked as answer]
https://react-hook-form.com/api/usefieldarray ?

**Reply by karlkras** [reply]
Hmm, maybe, but I'd prefer a controlled option.
I think I'm good rolling my own. Lotta work.
